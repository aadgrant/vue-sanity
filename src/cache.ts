import Vue from 'vue'
import CompositionApi, {
  Ref,
  ref,
  reactive,
  watch,
  computed,
  onServerPrefetch,
  getCurrentInstance,
} from '@vue/composition-api'

Vue.use(CompositionApi)

const cache = reactive<Record<string, any>>({})

export function ensureInstance() {
  const instance = getCurrentInstance()
  if (!instance) throw new Error('You must call this from within a component')
  return instance
}

export type FetchStatus =
  | 'server loaded'
  | 'loading'
  | 'client loaded'
  | 'error'

export interface CacheOptions<K> {
  initialValue?: K
  /**
   * Whether to disable SSR fetching. Defaults to false. Now replaced with strategy: 'client'
   * @deprecated
   */
  clientOnly?: boolean
  /**
   * Strategy for fetching. Defaults to 'both'.
   * 'server' will not refetch if the cache has been populated on SSR.
   * 'client' will disable SSR fetching.
   * 'both' will fetch on server and refetch when page is loaded.
   */
  strategy?: 'server' | 'client' | 'both'
}

export function useCache<T, K = null>(
  key: Ref<string>,
  fetcher: (key: string) => Promise<T>,
  options: CacheOptions<K> = {}
) {
  const instance = ensureInstance()
  const isServer = instance.$isServer

  const status = ref<FetchStatus>('loading')

  let { initialValue } = options

  const enableSSR =
    !options.clientOnly && (!options.strategy || options.strategy !== 'client')

  if (enableSSR && !isServer) {
    const prefetchState =
      (window as any).__VSANITY_STATE__ ||
      ((window as any).__NUXT__ && (window as any).__NUXT__.vsanity)
    if (prefetchState && prefetchState[key.value]) {
      initialValue = prefetchState[key.value]
      status.value = 'server loaded'
    }
  }

  function initialiseKey(key: string) {
    if (!cache[key]) Vue.set(cache, key, initialValue || null)
  }

  function setCache(key: string, value: any) {
    Vue.set(cache, key, value)
  }

  const error = ref(null)
  async function fetch(query = key.value) {
    try {
      setCache(query, (await fetcher(query)) || initialValue || null)
      status.value = isServer ? 'server loaded' : 'client loaded'
    } catch (e) {
      error.value = e
      status.value = 'error'
    }
  }

  if (enableSSR && isServer) {
    if (instance.$ssrContext) {
      if (instance.$ssrContext.nuxt && !instance.$ssrContext.nuxt.vsanity) {
        instance.$ssrContext.nuxt.vsanity = {}
      } else if (!instance.$ssrContext.vsanity) {
        instance.$ssrContext.vsanity = {}
      }
    }

    onServerPrefetch(async () => {
      await fetch()
      if (instance.$ssrContext) {
        if (instance.$ssrContext.nuxt) {
          instance.$ssrContext.nuxt.vsanity[key.value] = cache[key.value]
        } else {
          instance.$ssrContext.vsanity[key.value] = cache[key.value]
        }
      }
    })
  }

  watch(key, async key => {
    if (
      options.strategy === 'server' &&
      status.value === 'server loaded' &&
      ![null, initialValue].includes(cache[key])
    )
      return

    await fetch(key)
  })

  const data = computed(() => {
    initialiseKey(key.value)
    return (cache[key.value] || initialValue || null) as T | K
  })

  return {
    setCache,
    triggerFetch: fetch,
    fetch,
    data,
    status,
    error,
  }
}
