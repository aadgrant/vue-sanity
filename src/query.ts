import {
  computed,
  inject,
  watch,
  InjectionKey,
  Ref,
} from '@vue/composition-api'
import minifier from 'minify-groq'
import { SanityClient } from '@sanity/client'

import { useCache, CacheOptions, FetchStatus } from './cache'

export const clientSymbol: InjectionKey<SanityClient> = Symbol('Sanity client')
export const previewClientSymbol: InjectionKey<SanityClient> = Symbol(
  'Sanity client for previews'
)

type Query = () => string

interface Result<T> {
  /**
   * An automatically synced and updated result of the Sanity query.
   */
  data: Ref<T>
  /**
   * The status of the query. Can be 'server loaded', 'loading', 'client loaded' or 'error'.
   */
  status: Ref<FetchStatus>
}

type Options = Omit<CacheOptions<any>, 'initialValue'> & {
  /**
   * Whether to listen to real-time updates from Sanity. You can also pass an object of options to pass to `client.listen`. Defaults to false.
   */
  listen?: boolean | Record<string, any>
}

/**
 *
 * @param query A function that retuns a query string. If the return value changes, a new Sanity query will be run and the return value automatically updated.
 */
export function useSanityFetcher<T extends any>(query: Query): Result<T | null>

/**
 *
 * @param query A function that retuns a query string. If the return value changes, a new Sanity query will be run and the return value automatically updated.
 * @param initialValue The value to return before the Sanity client returns an actual result. Defaults to null.
 * @param mapper A function that transforms the result from Sanity, before returning it to your component.
 */
export function useSanityFetcher<T extends any, R extends any = T>(
  query: Query,
  initialValue: R,
  mapper?: (result: any) => T,
  options?: Options
): Result<T | R>

export function useSanityFetcher(
  query: Query,
  initialValue = null,
  mapper = (result: any) => result,
  options?: Options
) {
  const client = inject(clientSymbol)
  if (!client)
    throw new Error(
      'You must call useSanityClient before using sanity resources in this project.'
    )

  const computedQuery = computed(() => minifier(query()).replace(/\n/g, ' '))

  const { data, status, setCache } = useCache(
    computedQuery,
    query => client.fetch(query).then(mapper),
    {
      initialValue,
      ...options,
    }
  )

  if (options && options.listen) {
    const previewClient = inject(previewClientSymbol) || client

    const listenOptions =
      typeof options.listen === 'boolean' ? undefined : options.listen

    watch(computedQuery, query => {
      const subscription = previewClient
        .listen(query, listenOptions)
        .subscribe(event => setCache(query, event.result))

      const unwatch = watch(computedQuery, newQuery => {
        if (newQuery !== query) {
          subscription.unsubscribe()
          unwatch()
        }
      })
    })
  }

  return { data, status }
}
