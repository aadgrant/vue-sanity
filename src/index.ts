import Vue from 'vue'
import CompositionApi, { provide } from '@vue/composition-api'

import sanityClient, { ClientConfig } from '@sanity/client'
import imageUrlBuilder from '@sanity/image-url'

import { useCache, ensureInstance } from './cache'
import { useSanityImage, imageBuilderSymbol } from './image'
import { useSanityFetcher, clientSymbol, previewClientSymbol } from './query'

Vue.use(CompositionApi)

interface RequiredConfig {
  /**
   * Your project ID. You can find it in your sanity.json.
   */
  projectId: string
  /**
   * You must specify which dataset to use. You can find it in your sanity.json.
   */
  dataset: string
}

/**
 *
 * @param supportPreview Whether to create a preview client (that won't use CDN, and supports credentials for viewing drafts). Defaults to false.
 */
export function useSanityClient(
  config: ClientConfig & RequiredConfig,
  supportPreview = false
) {
  ensureInstance()

  const client = sanityClient(config)
  const imageBuilder = imageUrlBuilder(config)

  provide(clientSymbol, client)
  provide(imageBuilderSymbol, imageBuilder)

  if (supportPreview) {
    const previewClient = sanityClient({
      ...config,
      useCdn: false,
      token: undefined,
      withCredentials: true,
    })
    provide(previewClientSymbol, previewClient)
  }
}

export { useCache, useSanityFetcher, useSanityImage }
