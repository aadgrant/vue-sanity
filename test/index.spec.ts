import Vue from 'vue'
import CompositionApi, { ref } from '@vue/composition-api'
import flushPromises from 'flush-promises'

import { useSanityFetcher, useSanityClient, useSanityImage } from '../src'
import { runInSetup } from './helpers/mount'

Vue.config.productionTip = false
Vue.config.devtools = false

Vue.use(CompositionApi)

const mockFetch = jest.fn(async (key: string) => `return value-${key}`)
;(global.console.error as any) = jest
  .fn
  // console.info
  ()

const mockUnsubscribe = jest.fn()
const mockSubscribe = jest.fn((callback: (result: any) => void) => {
  callback({ result: 'sub update' })
  return {
    unsubscribe: mockUnsubscribe,
  }
})
const mockListen = jest.fn(() => ({
  subscribe: mockSubscribe,
}))

jest.mock('@sanity/client', () => {
  return jest.fn().mockImplementation(() => {
    return { fetch: mockFetch, listen: mockListen }
  })
})
// eslint-disable-next-line
const sanityClient = require('@sanity/client')

const config = {
  projectId: 'id',
  dataset: 'production',
}

beforeEach(() => {
  sanityClient.mockClear()
  ;(global.console.error as any).mockClear()
  mockListen.mockClear()
  mockSubscribe.mockClear()
  mockUnsubscribe.mockClear()
  mockFetch.mockClear()
})

describe('instantiator', () => {
  test('creates new sanity client', async () => {
    await runInSetup(() => {
      useSanityClient(config, true)
      return {}
    })

    expect(sanityClient).toHaveBeenCalledTimes(2)
    expect(sanityClient).toHaveBeenCalledWith(config)
    expect(sanityClient).toHaveBeenLastCalledWith({
      ...config,
      useCdn: false,
      token: undefined,
      withCredentials: true,
    })
  })
})

describe('image builder', () => {
  const image = {
    url: 'https://cdn.sanity.io/images/siteid/production/imageid-100x100.jpg',
    dimensions: {
      aspectRatio: 3,
      height: 3,
      width: 3,
    },
  }
  test('errors without proper config', async () => {
    await runInSetup(() => {
      useSanityImage(ref(image))
      return {}
    })
    // eslint-disable-next-line
    expect(console.error).toBeCalled()
  })

  test('errors when run outside of setup', async () => {
    let error
    try {
      useSanityImage(ref(image))
    } catch (e) {
      error = e
    }
    expect(error).toBeDefined()
  })

  test('produces relevant image url', async () => {
    const data = await runInSetup(() => {
      useSanityClient(config, true)
      const result = useSanityImage(ref(image))
      return { result }
    })

    const result = data.value.result

    expect(result).toMatchSnapshot()
  })

  test('works without dimensions', async () => {
    const data = await runInSetup(() => {
      useSanityClient(config, true)
      const result = useSanityImage(ref({ ...image, dimensions: undefined }))
      return { result }
    })

    const result = data.value.result

    expect(result).toMatchSnapshot()
  })
})

describe('fetcher', () => {
  test('errors when run outside of setup', async () => {
    let error
    try {
      useSanityFetcher(() => `outside-of-setup`)
    } catch (e) {
      error = e
    }
    expect(error).toBeDefined()
  })
  test('errors when client is not injected', async () => {
    await runInSetup(() => {
      useSanityFetcher(() => `my-error`)
      return {}
    })
    // eslint-disable-next-line
    expect(console.error).toBeCalled()
  })
  test('fetches query when slug updates', async () => {
    const slug = ref('key')

    await runInSetup(() => {
      useSanityClient(config)
      useSanityFetcher(() => `my-${slug.value}`)

      return {}
    })
    expect(mockFetch).toHaveBeenCalledWith(`my-key`)

    slug.value = 'new-key'
    await flushPromises()
    expect(mockFetch).toHaveBeenCalledWith(`my-new-key`)
  })

  test('status updates correctly', async () => {
    const data = await runInSetup(() => {
      useSanityClient(config)

      const { status } = useSanityFetcher(() => `my-key-status`, 'apple')
      return { status }
    })
    expect(data.value.status).toEqual('client loaded')
  })

  test('data updates correctly', async () => {
    const data = await runInSetup(() => {
      useSanityClient(config)

      const { data } = useSanityFetcher(() => `my-key-data`, 'apple')
      return { data }
    })
    expect(data.value.data).toEqual('return value-my-key-data')
  })

  test('subscribes to a sanity resource', async () => {
    const key = ref('subscription')
    await runInSetup(() => {
      useSanityClient(config)

      const { data } = useSanityFetcher(
        () => `my-key-${key.value}`,
        'apple',
        q => q,
        { listen: true }
      )
      return { data }
    })
    expect(mockSubscribe).toHaveBeenCalled()
    expect(mockListen).toHaveBeenCalled()

    key.value = 'new-sub'
    await flushPromises()
    expect(mockUnsubscribe).toHaveBeenCalled()
  })

  test('subscribes to a sanity resource with preview client', async () => {
    const key = ref('preview-subscription')
    await runInSetup(() => {
      useSanityClient(config, true)

      const { data } = useSanityFetcher(
        () => `my-key-${key.value}`,
        'apple',
        q => q,
        { listen: true }
      )
      return { data }
    })
    expect(mockSubscribe).toHaveBeenCalled()
    expect(mockListen).toHaveBeenCalled()

    key.value = 'new-sub'
    await flushPromises()
    expect(mockUnsubscribe).toHaveBeenCalled()
  })

  test('passes relevant options to listener', async () => {
    const key = ref('listen-options')
    const listenOptions = { sampleId: 30 }
    await runInSetup(() => {
      useSanityClient(config, true)

      const { data } = useSanityFetcher(
        () => `my-key-${key.value}`,
        'apple',
        q => q,
        { listen: listenOptions }
      )
      return { data }
    })
    expect(mockListen).toHaveBeenCalledWith(
      'my-key-listen-options',
      listenOptions
    )
  })
})
