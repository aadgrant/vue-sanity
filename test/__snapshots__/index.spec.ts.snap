// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`image builder produces relevant image url 1`] = `
Object {
  "placeholder": "",
  "sizes": "(max-width: 400px) 300px, (max-width: 700px) 600px, (max-width: 1300px) 1200px, (max-width: 2020px) 1920px, 3px",
  "src": "https://cdn.sanity.io/images/siteid/production/imageid-100x100.jpg",
  "srcset": "https://cdn.sanity.io/images/id/production/imageid-100x100.jpg?rect=0,33,100,33&w=300&h=100&q=82&fit=min 300w, https://cdn.sanity.io/images/id/production/imageid-100x100.jpg?rect=0,33,100,33&w=600&h=200&q=82&fit=min 600w, https://cdn.sanity.io/images/id/production/imageid-100x100.jpg?rect=0,33,100,33&w=1200&h=400&q=82&fit=min 1200w, https://cdn.sanity.io/images/id/production/imageid-100x100.jpg?rect=0,33,100,33&w=1920&h=640&q=82&fit=min 1920w, https://cdn.sanity.io/images/siteid/production/imageid-100x100.jpg 3w",
}
`;

exports[`image builder works without dimensions 1`] = `
Object {
  "src": "https://cdn.sanity.io/images/siteid/production/imageid-100x100.jpg",
}
`;
